csharp/src/Ice/Communicator.cs

```diff
@@ -360,7 +373,126 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
 
                 TraceLevels = new TraceLevels(this);
 
-                DefaultsAndOverrides = new DefaultsAndOverrides(this, Logger);
+                DefaultHost = GetProperty("Ice.Default.Host");
+
+                if (GetProperty("Ice.Default.SourceAddress") is string address)
+                {
+                    DefaultSourceAddress = Network.GetNumericAddress(address);
+                    if (DefaultSourceAddress == null)
+                    {
+                        throw new InvalidConfigurationException(
+                            $"invalid IP address set for Ice.Default.SourceAddress: `{address}'");
+                    }
+                }
+
+                DefaultTransport = GetProperty("Ice.Default.Transport") ?? "tcp";
+
+                DefaultCollocationOptimization = (GetPropertyAsInt("Ice.Default.CollocationOptimized") ?? 1) > 0;
+
+                string val = GetProperty("Ice.Default.EndpointSelection") ?? "Random";
+                if (val.Equals("Random"))
+                {
+                    DefaultEndpointSelection = Ice.EndpointSelectionType.Random;
+                }
+                else if (val.Equals("Ordered"))
+                {
+                    DefaultEndpointSelection = EndpointSelectionType.Ordered;
+                }
+                else
+                {
+                    throw new InvalidConfigurationException($"illegal value `{val}'; expected `Random' or `Ordered'");
+                }
+
+                DefaultTimeout = GetPropertyAsInt("Ice.Default.Timeout") ?? 60000;
+                if (DefaultTimeout < 1 && DefaultTimeout != -1)
+                {
+                    Logger.Warning($"invalid value for Ice.Default.Timeout `{DefaultTimeout}': defaulting to 60000");
+                    DefaultTimeout = 60000;
+                }
+
+                DefaultLocatorCacheTimeout = GetPropertyAsInt("Ice.Default.LocatorCacheTimeout") ?? -1;
+                if (DefaultLocatorCacheTimeout < -1)
+                {
+                    Logger.Warning(
+                    $"invalid value for Ice.Default.LocatorCacheTimeout `{DefaultLocatorCacheTimeout}': defaulting to -1");
```

Comment:
```suggestion
                    Logger.Warning(@$"invalid value for Ice.Default.LocatorCacheTimeout `{DefaultLocatorCacheTimeout
                        }': defaulting to -1");
```

Reply:
I didn't know you could use verbatim strings like that... Pretty cool!